File: web/tests/geo.test.ts
import { describe, it, expect } from 'vitest'
import {
  speedKmh,
  headingCardinal,
  locationContext,
  locationMetadata,
  mergeLocationMeta,
  type GeoFix,
} from '@/lib/geo'

/**
 * 📍 lib/geo — the pure half of the geolocation plumbing: unit conversion,
 * compass points, and the agent-facing `### Location` block (agi-diy's
 * context grammar, shared across web/iOS/Android).
 */

const fix = (over: Partial<GeoFix> = {}): GeoFix => ({
  lat: 37.7749,
  lng: -122.4194,
  accuracy: 12,
  altitude: 52.4,
  speed: 6.5,
  heading: 48,
  timestamp: 1753400000000,
  ...over,
})

describe('speedKmh', () => {
  it('converts m/s to km/h with one decimal', () => {
    expect(speedKmh(6.5)).toBe(23.4)
    expect(speedKmh(0)).toBe(0)
  })

  it('passes null through and rejects junk', () => {
    expect(speedKmh(null)).toBeNull()
    expect(speedKmh(NaN)).toBeNull()
    expect(speedKmh(-1)).toBeNull()
  })
})

describe('headingCardinal', () => {
  it('maps degrees to the eight compass points', () => {
    expect(headingCardinal(0)).toBe('N')
    expect(headingCardinal(48)).toBe('NE')
    expect(headingCardinal(90)).toBe('E')
    expect(headingCardinal(180)).toBe('S')
    expect(headingCardinal(270)).toBe('W')
    expect(headingCardinal(315)).toBe('NW')
    // seam: 337.5 sits exactly between NW and N — rounds toward N
    expect(headingCardinal(337.5)).toBe('N')
  })

  it('wraps: 359° is N again, negatives normalize', () => {
    expect(headingCardinal(359)).toBe('N')
    expect(headingCardinal(720 + 90)).toBe('E')
    expect(headingCardinal(-90)).toBe('W')
  })

  it('null/NaN → null (browser sends null when not moving)', () => {
    expect(headingCardinal(null)).toBeNull()
    expect(headingCardinal(NaN)).toBeNull()
  })
})

describe('locationContext', () => {
  it('renders the full agi-diy block for a moving fix', () => {
    expect(locationContext(fix())).toBe(
      [
        '### Location',
        '- **Coordinates**: 37.7749, -122.4194',
        '- **Accuracy**: ±12m',
        '- **Altitude**: 52m',
        '- **Speed**: 23.4 km/h',
        '- **Heading**: NE (48°)',
      ].join('\n'),
    )
  })

  it('omits speed/heading when stationary (browser nulls)', () => {
    const block = locationContext(fix({ speed: null, heading: null, altitude: null }))
    expect(block).toContain('- **Coordinates**: 37.7749, -122.4194')
    expect(block).not.toContain('Speed')
    expect(block).not.toContain('Heading')
    expect(block).not.toContain('Altitude')
  })

  it('omits a zero speed line — "0.0 km/h" is noise, absence means parked', () => {
    expect(locationContext(fix({ speed: 0 }))).not.toContain('Speed')
  })

  it('returns empty string for no/degenerate fix so callers append blindly', () => {
    expect(locationContext(null)).toBe('')
    expect(locationContext(fix({ lat: NaN }))).toBe('')
  })
})

describe('locationMetadata', () => {
  it('coarsens coords to 4 decimals and carries derived units', () => {
    expect(locationMetadata(fix({ lat: 37.77491234, lng: -122.41945678 }))).toEqual({
      lat: 37.7749,
      lng: -122.4195,
      accuracyM: 12,
      speedKmh: 23.4,
      heading: 'NE',
    })
  })

  it('drops absent fields instead of sending nulls', () => {
    expect(locationMetadata(fix({ accuracy: null, speed: null, heading: null }))).toEqual({
      lat: 37.7749,
      lng: -122.4194,
    })
  })

  it('null fix → null', () => {
    expect(locationMetadata(null)).toBeNull()
  })
})

describe('mergeLocationMeta', () => {
  const moving = fix()

  it('no fix → metadata passes through byte-identical (header unchanged)', () => {
    expect(mergeLocationMeta('.', null)).toBe('.')
    expect(mergeLocationMeta('', null)).toBe('')
    const weather = { tempC: 21 }
    expect(mergeLocationMeta(weather, null)).toBe(weather)
  })

  it('object metadata (404 weather) gains a location key', () => {
    const merged = mergeLocationMeta({ tempC: 21 }, moving) as any
    expect(merged.tempC).toBe(21)
    expect(merged.location).toEqual(locationMetadata(moving))
  })

  it("scalar metadata ('.' home tiny) wraps beside the location", () => {
    expect(mergeLocationMeta('.', moving)).toEqual({
      note: '.',
      location: locationMetadata(moving),
    })
  })

  it("empty metadata ('' slug pages) becomes just the location", () => {
    expect(mergeLocationMeta('', moving)).toEqual({ location: locationMetadata(moving) })
  })
})

File: android/app/src/test/java/technology/tiny/app/geo/GeoTest.kt
package technology.tiny.app.geo

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * 📍 Geo's pure half — must match web lib/geo.ts (tests/geo.test.ts)
 * byte-for-byte: every client teaches the tiny the same `### Location`
 * shape, so a drift here is a cross-platform context fork.
 */
class GeoTest {
    private fun fix(
        speedMs: Double? = 6.5,
        headingDeg: Double? = 48.0,
        accuracyM: Int? = 12,
        altitudeM: Int? = 52,
    ) = Geo.Fix(
        lat = 37.7749, lng = -122.4194,
        accuracyM = accuracyM, altitudeM = altitudeM,
        speedMs = speedMs, headingDeg = headingDeg,
        timestampMs = 1_753_400_000_000,
    )

    @Test fun `kmh converts with one decimal and rejects junk`() {
        assertEquals(23.4, Geo.kmh(6.5)!!, 1e-9)
        assertEquals(0.0, Geo.kmh(0.0)!!, 1e-9)
        assertNull(Geo.kmh(null))
        assertNull(Geo.kmh(-1.0))
        assertNull(Geo.kmh(Double.NaN))
    }

    @Test fun `cardinal maps and wraps like the web`() {
        assertEquals("N", Geo.cardinal(0.0))
        assertEquals("NE", Geo.cardinal(48.0))
        assertEquals("E", Geo.cardinal(90.0))
        assertEquals("S", Geo.cardinal(180.0))
        assertEquals("W", Geo.cardinal(270.0))
        assertEquals("NW", Geo.cardinal(315.0))
        assertEquals("N", Geo.cardinal(359.0))
        assertEquals("E", Geo.cardinal(810.0))
        assertEquals("W", Geo.cardinal(-90.0))
        assertNull(Geo.cardinal(null))
        assertNull(Geo.cardinal(Double.NaN))
    }

    @Test fun `contextBlock renders the exact web grammar for a moving fix`() {
        assertEquals(
            listOf(
                "### Location",
                "- **Coordinates**: 37.7749, -122.4194",
                "- **Accuracy**: ±12m",
                "- **Altitude**: 52m",
                "- **Speed**: 23.4 km/h",
                "- **Heading**: NE (48°)",
            ).joinToString("\n"),
            Geo.contextBlock(fix()),
        )
    }

    @Test fun `stationary fix omits speed and heading lines`() {
        val block = Geo.contextBlock(fix(speedMs = null, headingDeg = null, altitudeM = null))
        assertEquals(
            listOf(
                "### Location",
                "- **Coordinates**: 37.7749, -122.4194",
                "- **Accuracy**: ±12m",
            ).joinToString("\n"),
            block,
        )
    }

    @Test fun `zero speed is parked, not a 0 kmh line`() {
        val block = Geo.contextBlock(fix(speedMs = 0.0, headingDeg = null))
        assertEquals(false, block.contains("Speed"))
    }

    @Test fun `null or degenerate fix renders empty so callers append blindly`() {
        assertEquals("", Geo.contextBlock(null))
        assertEquals("", Geo.contextBlock(fix().copy(lat = Double.NaN)))
    }
}
